src/builtins.c

```diff
@@ -1961,6 +1969,7 @@ void jl_init_primitives(void) JL_GC_DISABLED
     jl_builtin__typebody = add_builtin_func("_typebody!", jl_f__typebody);
     add_builtin_func("_equiv_typedef", jl_f__equiv_typedef);
     jl_builtin_donotdelete = add_builtin_func("donotdelete", jl_f_donotdelete);
+    add_builtin_func("_add_finalizer", jl_f__add_finalizer);
```

Comment:
Is there a reason this needs to have a weird internal name, and not simply `Core.finalizer`?

Reply:
I would also argue for splitting this PR into a couple commits (changing `ccall` -> `builtin` with some codegen support; adding the new effect type (notls); and the key optimization here) so that if someone later has an issue, they can point to which part of the change was a problem for them in bisection. Keeping this as one PR seems fine, just merging it as a few related commits.